### PR TITLE
Tin 9 Show list of friends sorted by rating

### PR DESCRIPTION
**Pull request by Kr1s4an**
Screenshot from 2024-02-05 10-38-00
Screenshot from 2024-02-05 10-38-41

### CODE REVIEW

File: src/main/java/com/volasoftware/tinder/controller/FriendController.java [hunk 1]
```diff
@@ -39,7 +41,7 @@ public ResponseEntity<String> seedFriends(@RequestParam(required = false) Long u
 
     @GetMapping("/")
     @SecurityRequirement(name = "Bearer Authentication")
-    public ResponseEntity<List<FriendDetails>> getFriendsSortedByLocation(@RequestBody FriendSearchDto friendSearchDto) {
+    public ResponseEntity<List<FriendDetails>> getFriendsSortedByLocation(FriendSearchDto friendSearchDto) {
```
Comment (Tsvetoslav-Garkov) [inline src/main/java/com/volasoftware/tinder/controller/FriendController.java:44]: Why did you remove `@RequestBody` from the FriendSearchDto?

File: src/main/java/com/volasoftware/tinder/controller/FriendController.java [hunk 2]
```diff
@@ -59,4 +61,15 @@ public ResponseEntity rateFriends(@RequestBody RatingDto ratingDto) {
 
         return ResponseEntity.ok(HttpStatus.OK);
     }
+
+    @GetMapping("/rating")
+    @SecurityRequirement(name = "Bearer Authentication")
+    public ResponseEntity<Page<FriendRatingDetails>> getFriendsSortedByRating(
+            @RequestParam(defaultValue = "0") int page,
```
Comment (Tsvetoslav-Garkov) [inline src/main/java/com/volasoftware/tinder/controller/FriendController.java]: Its a good idea that you added defaultValue to the request params, but also add `required = false`, just to make it available to call the endpoint without request params. Because now I won't be able to call it when I do not provide a request param.

File: src/main/java/com/volasoftware/tinder/repository/UserRepository.java [hunk 3]
```diff
@@ -43,4 +46,14 @@ public interface UserRepository extends JpaRepository<User, java.lang.Long> {
     List<FriendDetails> findUserFriendsSortedByLocation(@Param("id") Long id,
                                                         @Param("lat") Double lat,
                                                         @Param("lon") Double lon);
+
+    @Query(value = "SELECT u.first_name as firstName, u.last_name as lastName, u.age as age, r.rating as rating " +
+            " FROM user u " +
+            " LEFT JOIN rating r ON u.id = r.friend_id AND r.user_id = :id " +
+            " WHERE u.id IN (SELECT friend_id FROM friend WHERE user_id = :id) " +
+            " ORDER BY rating DESC",
+            countQuery = "SELECT COUNT(u.id) FROM user u WHERE u.id IN " +
```
Comment (Tsvetoslav-Garkov) [inline src/main/java/com/volasoftware/tinder/repository/UserRepository.java:54]: You can remove the ORDER BY from the query because in the service code I saw you add an Sort object into the supplied pageable. Its redundant.

File: src/test/java/com/volasoftware/tinder/UserServiceTest.java [hunk 4]
```diff
@@ -414,4 +418,102 @@ public void testFindFriendByIdWithNotFriends() {
...
+        verify(userRepository, times(1)).findFriendsSortedByRating(loggedUser.getId(), pageRequest);
+        verifyNoMoreInteractions(userRepository);
+    }
+
+    @Test
+    public void testFindFriendsSortedByRatingWithMultipleFriends() {
+        User loggedUser = new User();
+        loggedUser.setId(1L);
+
+        FriendRatingDetailsImpl mockFriend1 = new FriendRatingDetailsImpl("John", "Doe", 25, 10);
```
Comment (Tsvetoslav-Garkov) [inline src/test/java/com/volasoftware/tinder/UserServiceTest.java:487]: Just have seen this, can you test this method with different user data, not just for each to be John Doe 25 10, what will happed if we change one of the mocked friends?

Comment (Kr1s4an) [reply src/test/java/com/volasoftware/tinder/UserServiceTest.java:487]: Thanks for reminding me. I forgot about it. I was meant to do it.

**Review by Tsvetoslav-Garkov — APPROVED** [review]
No more comments from me, all issues pointed are resolved.
